File: leetcode/39-combination-sum/main.py
"""
    questions to ask:
    - will there be zero? no
    - will there to duplicate numbers? no
"""


class Solution(object):
    """
    1st attempt:
    - iterative bottom-up dynamic programming approch
    - use a hashtable to avoid duplicates
    e.g. candidates = [2,3,5], target = 8
    f[0] = []
    f[1] = []
    f[2] = [[2]]
    f[3] = [[3]]
    f[4] = [[2,2]]
    f[5] = [[2,3],[5]]
    f[6] = [[2,2,2],[3,3]]
    f[7] = [[2,2,3],[2,5]]
    f[8] = [[2, 2, 2, 2], [2, 3, 3], [3, 5]]
    beats 15.56%
    """

    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        result = []
        for i in range(target+1):
            result.append([])
        for i in range(target+1):
            future_result = []
            seen = {}
            for candidate in candidates:
                cur = i - candidate
                if cur < 0:
                    continue
                that_results = result[cur]
                if len(that_results) == 0 and cur == 0:
                    future_result.append([candidate])
                else:
                    for that_result in that_results:
                        temp = that_result + [candidate]
                        temp = sorted(temp)
                        key = "".join(str(n) for n in temp)
                        if key in seen:
                            continue
                        seen[key] = True
                        future_result.append(temp)

            result[i] = future_result

        return result[target]


s = Solution()

print(s.combinationSum([2, 3, 5], 8))
print(s.combinationSum([2, 3, 6, 7], 7))
print(s.combinationSum([1, 2, 3, 4], 15))
print("------------------------------")


class Solution(object):
    """
    2nd approach: recursive dfs, avoid duplicate by considering the candidates which are >= num
    beats 76.61%
    """

    def __init__(self):
        self.result = []

    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """
        candidates.sort()
        self.dfs(candidates, target, [], 0)
        return self.result

    def dfs(self, candidates, target, path, from_idx):
        if target == 0:
            self.result.append(path)
            return
        for i in range(from_idx, len(candidates)):
            if target < candidates[i]:
                return
            can = candidates[i]
            self.dfs(candidates, target-can, path+[can], i)


s = Solution()

print(s.combinationSum([2, 3, 5], 8))
print(s.combinationSum([2, 3, 6, 7], 7))
print(s.combinationSum([1, 2, 3, 4], 15))
print("------------------------------")


class Solution(object):
    """
    2nd approach: recursive dfs, avoid duplicate by considering the candidates which are >= num
    144 ms, faster than 14.76%
    """
    def combinationSum(self, candidates, target):
        candidates.sort()
        self.res = []
        self.dfs(candidates, [], target)
        return self.res

    def dfs(self, candidates, chosen, target):
        if target == 0:
            self.res.append(chosen)
        elif target > 0:
            for i in range(len(candidates)):
                x = candidates[i]
                if len(chosen) == 0 or chosen[-1] <= x:
                    self.dfs(candidates, chosen + [x], target - x)


s = Solution()

print(s.combinationSum([2, 3, 5], 8))
print(s.combinationSum([2, 3, 6, 7], 7))
print(s.combinationSum([1, 2, 3, 4], 15))
print("------------------------------")

File: leetcode/39-combination-sum/main.go
package main

import (
	"fmt"
	"sort"
	"strconv"
)

/*
naive approach: top-down recursion
TLE because there are too many duplicate calculations
func combinationSum(candidates []int, target int) [][]int {
	result := [][]int{}
	seen := make(map[string]bool)
	var dfs func(num int, path []int)
	dfs = func(num int, path []int) {
		if num == 0 {
			sort.Ints(path)
			key := ints2str(path)
			if _, x := seen[key]; !x {
				seen[key] = true
				result = append(result, path)
			}
		} else if num > 0 {
			for i := 0; i < len(candidates); i++ {
				cur := candidates[i]
				nextPath := []int{}
				nextPath = append(nextPath, path...)
				nextPath = append(nextPath, cur)
				if num < cur {
					break
				}
				dfs(num-cur, nextPath)
			}
		}
	}
	dfs(target, []int{})
	return result
}
*/

/*
	1st attempt:
	- bottom-up dynamic programming approch
	- use a hashtable to avoid duplicates
	e.g. candidates = [2,3,5], target = 8
	f[0] = []
	f[1] = []
	f[2] = [[2]]
	f[3] = [[3]]
	f[4] = [[2,2]]
	f[5] = [[2,3],[5]]
	f[6] = [[2,2,2],[3,3]]
	f[7] = [[2,2,3],[2,5]]
	f[8] = [[2, 2, 2, 2], [2, 3, 3], [3, 5]]
	beats 23.85%
*/
func combinationSum(candidates []int, target int) [][]int {
	result := [][][]int{}
	for i := 0; i <= target; i++ {
		result = append(result, [][]int{})
	}
	for i := 0; i <= target; i++ {
		futureResult := [][]int{}
		seen := make(map[string]bool)
		for j := 0; j < len(candidates); j++ {
			candidate := candidates[j]
			remain := i - candidate
			if remain < 0 {
				continue
			}
			remainResults := result[remain]
			if len(remainResults) == 0 && remain == 0 {
				futureResult = append(futureResult, []int{candidate})
			} else {
				for k := 0; k < len(remainResults); k++ {
					temp := []int{}
					temp = append(temp, remainResults[k]...)
					temp = append(temp, candidate)
					sort.Ints(temp)
					key := ints2str(temp)
					if _, x := seen[key]; x {
						continue
					}
					seen[key] = true
					futureResult = append(futureResult, temp)
				}
			}
		}
		result[i] = futureResult
	}
	return result[target]
}

func ints2str(nums []int) string {
	result := ""
	for i := 0; i < len(nums); i++ {
		result += strconv.Itoa(nums[i])
	}
	return result
}

func main() {
	fmt.Println(combinationSum([]int{2, 3, 5}, 8))
	fmt.Println(combinationSum([]int{2, 3, 6, 7}, 7))
	fmt.Println(combinationSum([]int{1, 2, 3, 4}, 15))
}
